server.go

```diff
@@ -4173,13 +4327,297 @@ func (s *server) SubscribeCustomMessages() (*subscribe.Client, error) {
+		return nil
+
+	case temporary:
+		// If this peer is temporary, we need to check if it will revert
+		// to a restricted-access peer.
+		numPending, pendingFound := s.pendingChannelPeers[remoteHex]
+		if !pendingFound {
+			s.banScoreMtx.Unlock()
+
+			// Error if we did not find any info in
+			// pendingChannelPeers.
+			return ErrNoPendingPeerInfo
+		}
+
+		currentNumPending := numPending - 1
+
+		if currentNumPending == 0 {
+			// Remove the entry from pendingChannelPeers.
+			delete(s.pendingChannelPeers, remoteHex)
+
+			// If this is the only pending-open channel for this
+			// peer and it's getting removed, attempt to demote
+			// this peer to a restricted peer.
+			if s.numRestricted == maxRestrictedSlots {
+				// There are no available restricted slots, so
+				// we need to disconnect this peer.
+				s.banScoreMtx.Unlock()
+
+				remotePub, err := btcec.ParsePubKey(
+					[]byte(remoteHex),
+				)
+				if err != nil {
+					return err
+				}
+
+				// The banScoreMtx cannot be held when the
+				// below DisconnectPeer call occurs.
+				return s.DisconnectPeer(remotePub)
+			}
+
+			// Otherwise, there is an available restricted-access
+			// slot, so we can demote this peer.
+			s.peerScores[remoteHex] = peerSlotStatus{
+				state: restricted,
+			}
+
+			// Update numRestricted.
+			s.numRestricted++
+
+			s.banScoreMtx.Unlock()
+
+			return nil
+		}
+
+		// Else, we don't need to demote this peer since it has other
+		// pending-open channels with us.
+		s.pendingChannelPeers[remoteHex] = currentNumPending
+
+		s.banScoreMtx.Unlock()
+
+		return nil
+
+	case restricted:
+		s.banScoreMtx.Unlock()
+
+		// This should not be possible. This indicates an error.
+		return fmt.Errorf("invalid peer access state transition")
+
+	default:
+		s.banScoreMtx.Unlock()
+
+		// This should not be possible.
+		return fmt.Errorf("invalid peer access status")
+	}
+}
+
+// NewOpenChan is called when a pending-open channel becomes an open channel
+// (i.e. the funding transaction has confirmed). If the remote peer is a
+// temporary-access peer, it will be promoted to a protected-access peer.
+func (s *server) NewOpenChan(remoteHex string) error {
+	s.banScoreMtx.Lock()
+	defer s.banScoreMtx.Unlock()
+
+	// Fetch the peer's access status from peerScores.
+	status, found := s.peerScores[remoteHex]
+	if !found {
+		// If we didn't find the peer, we'll return an error.
+		return ErrNoPeerScore
+	}
+
+	switch status.state {
+	case protected:
+		// If the peer's state is already protected, we don't need to
+		// do anything more.
+		return nil
+
+	case temporary:
+		// If the peer's state is temporary, we'll upgrade the peer to
+		// a protected peer. This will remove its entry from
+		// pendingChannelPeers and insert an entry into channelPeers.
+		delete(s.pendingChannelPeers, remoteHex)
+		s.channelPeers[remoteHex] = struct{}{}
+
+		newStatus := peerSlotStatus{
+			state: protected,
+		}
+		s.peerScores[remoteHex] = newStatus
```

Comment:
I don't think we really need the `peerScores` map either, we can just have a function returning the status based on channel state counts.

Reply:
It's not useful now, but the idea is to track a score based on other things as well

Reply:
prefer to keep as-is for now in case we actually start scoring